cdisc_rules_engine/check_operators/helpers.py

```diff
@@ -56,91 +93,210 @@ def is_valid_duration(duration: str, negative) -> bool:
+    return date_components, time_components, has_time
 
 
-def get_hour(date_string: str):
-    timestamp = get_date(date_string)
-    return timestamp.hour
+@lru_cache(maxsize=1000)
+def detect_datetime_precision(date_str: str) -> DatePrecision | None:
+    if not _datestring_is_valid(date_str):
+        return None
+    date_components, time_components, has_time = _parse_datetime_string(date_str)
+    if all(_check_date_component_missing(component) for component in date_components):
+        return None
+    return _date_and_time_precision(date_components, time_components, has_time)
 
 
-def get_minute(date_string: str):
-    timestamp = get_date(date_string)
-    return timestamp.minute
+def _datestring_is_valid(date_str: str) -> bool:
+    return bool(date_str and isinstance(date_str, str) and date_regex.match(date_str))
 
 
-def get_second(date_string: str):
-    timestamp = get_date(date_string)
-    return timestamp.second
+def _check_date_component_missing(component) -> bool:
+    return component is None or component == "-" or component == ""
 
 
-def get_microsecond(date_string: str):
-    timestamp = get_date(date_string)
-    return timestamp.microsecond
+def _get_precision_before(precision: DatePrecision) -> DatePrecision | None:
+    if precision == DatePrecision.year:
+        return None
+    prev_index = precision.value - 1
+    if prev_index >= 0:
+        return DatePrecision(prev_index)
+    return None
+
+
+def _check_date_precision(date_components) -> tuple:
+    for precision in [DatePrecision.year, DatePrecision.month, DatePrecision.day]:
+        component = (
+            date_components[precision.value]
+            if precision.value < len(date_components)
+            else None
+        )
+        if _check_date_component_missing(component):
+            result = _get_precision_before(precision)
+            return (True, result)
+    return (False, None)
+
+
+def _check_time_precision(time_components) -> DatePrecision:
+    if not time_components:
+        return DatePrecision.day
+    hour, minute, second, microsecond = time_components
+    if _check_date_component_missing(hour):
+        return DatePrecision.day
+    if _check_date_component_missing(minute):
+        return DatePrecision.hour
+    if second is None:
+        return DatePrecision.minute
+    if _check_date_component_missing(second):
+        return DatePrecision.minute
+    if microsecond is None:
+        return DatePrecision.second
+    if not microsecond or microsecond == "":
+        return DatePrecision.second
+    return DatePrecision.microsecond
+
+
+def _date_and_time_precision(
+    date_components, time_components, has_time
+) -> DatePrecision | None:
+    found_missing, date_result = _check_date_precision(date_components)
+    if found_missing:
+        return date_result
+    if not has_time or not time_components:
+        return DatePrecision.day
+    return _check_time_precision(time_components)
```

Comment:
Can you explain why these need to be separate? If they don't need to be separate, can they be combined to a single method? Remember to take advantage of iterating through enums like I've done elsewhere instead of handwriting each case.